src/librustdoc/clean/auto_trait.rs

```diff
@@ -99,29 +131,126 @@ impl<'a, 'tcx, 'rcx> AutoTraitFinder<'a, 'tcx, 'rcx> {
+        let mut traits = Vec::new();
+        if self.cx.crate_name != Some("core".to_string()) &&
+           self.cx.access_levels.borrow().is_doc_reachable(def_id) {
+            if let ty::TyAdt(_adt, _) = ty.sty {
+                let real_name = name.clone().map(|name| Ident::from_str(&name));
+                let param_env = self.cx.tcx.param_env(def_id);
+                for &trait_def_id in self.cx.all_traits.iter() {
+                    if !self.cx.access_levels.borrow().is_doc_reachable(trait_def_id) ||
+                       self.cx.generated_synthetics
+                              .borrow_mut()
+                              .get(&(def_id, trait_def_id))
+                              .is_some() {
+                        continue
+                    }
+                    self.cx.tcx.for_each_relevant_impl(trait_def_id, ty, |impl_def_id| {
+                        self.cx.tcx.infer_ctxt().enter(|infcx| {
+                            let generics = infcx.tcx.generics_of(impl_def_id);
+                            let trait_ref = infcx.tcx.impl_trait_ref(impl_def_id).unwrap();
+
+                            if !match infcx.tcx.type_of(impl_def_id).sty {
+                                ::rustc::ty::TypeVariants::TyParam(_) => true,
+                                _ => false,
+                            } {
+                                return
+                            }
+
+                            let substs = infcx.fresh_substs_for_item(DUMMY_SP, def_id);
+                            let ty2 = ty.subst(infcx.tcx, substs);
+                            let param_env = param_env.subst(infcx.tcx, substs);
+
+                            let impl_substs = infcx.fresh_substs_for_item(DUMMY_SP, impl_def_id);
+                            let trait_ref = trait_ref.subst(infcx.tcx, impl_substs);
+
+                            // Require the type the impl is implemented on to match
+                            // our type, and ignore the impl if there was a mismatch.
+                            let cause = traits::ObligationCause::dummy();
+                            let eq_result = infcx.at(&cause, param_env)
+                                                 .eq(trait_ref.self_ty(), ty2);
+                            if let Ok(InferOk { value: (), obligations }) = eq_result {
+                                // FIXME(eddyb) ignoring `obligations` might cause false positives.
+                                drop(obligations);
+
+                                let may_apply = infcx.predicate_may_hold(&traits::Obligation::new(
+                                    cause.clone(),
+                                    param_env,
+                                    trait_ref.to_predicate(),
+                                ));
+                                if !may_apply {
+                                    return
+                                }
+                                self.cx.generated_synthetics.borrow_mut()
```

Comment:
couldn't we do this way earlier (like where we're early aborting usually)

Reply:
No because a trait can have multiple generic impls and we only know at this point that the current impl matches.